src/ErrorProne.NET/ErrorProne.NET/Rules/ExceptionHandling/ArgumentExceptionUsesStringAnalyzer.cs

```diff
@@ -0,0 +1,88 @@
+
+        private const string ParamNameShouldNotBeStringRuleTitle = "Use nameof instead of string.";
+        private const string ParamNameShouldNotBeStringRuleMessageFormat = "ArgumentException constructor obtains parameter name from string \"{0}\" instead of nameof({0}).";
+        private const string ParamNameShouldNotBeStringRuleDescription = "It is better to obtain string name of the paramater by using nameof operator to avoid typos and post-refactoring errors.";
+
+        private const string Category = "CodeSmell";
+
+        private static readonly DiagnosticDescriptor ParamNameRequiredRule =
+            new DiagnosticDescriptor(RuleIds.ArgumentExceptionParamNameRequired, ParamNameRequiredTitle, ParamNameRequiredMessage, Category, DiagnosticSeverity.Warning, description: ParamNameRequiredDescription, isEnabledByDefault: true);
+        private static readonly DiagnosticDescriptor MethodHasNoSuchParamNameRule =
+            new DiagnosticDescriptor(RuleIds.ArgumentExceptionMethodHasNoSuchParamName, MethodHasNoSuchParamNameRuleTitle, MethodHasNoSuchParamNameRuleMessageFormat, Category, DiagnosticSeverity.Warning, description: MethodHasNoSuchParamNameRuleDescription, isEnabledByDefault: true);
+        private static readonly DiagnosticDescriptor ParamNameShouldNotBeStringRule =
+            new DiagnosticDescriptor(RuleIds.ArgumentExceptionParamNameShouldNotBeString, ParamNameShouldNotBeStringRuleTitle, ParamNameShouldNotBeStringRuleMessageFormat, Category, DiagnosticSeverity.Warning, description: ParamNameShouldNotBeStringRuleDescription, isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(ParamNameRequiredRule, MethodHasNoSuchParamNameRule, ParamNameShouldNotBeStringRule);
+
+        public override void Initialize(AnalysisContext context) {
+            context.RegisterSyntaxNodeAction(AnalyzeObjectCreation, SyntaxKind.ObjectCreationExpression);
+        }
+
+        private void AnalyzeObjectCreation(SyntaxNodeAnalysisContext context) {
+            var objectCreation = (ObjectCreationExpressionSyntax)context.Node;
+            var method = context.GetCtorSymbol(objectCreation);
+            if (!Equals(method.ContainingType, context.GetClrType<ArgumentException>())) {
+                return;
+            }
+
+            var stringType = context.GetClrType<string>();
+            var paramNameArgument = method.Parameters.FirstOrDefault(p => p.Name == "paramName");
+            if (paramNameArgument == null || !Equals(paramNameArgument.Type, stringType)) {
+                context.ReportDiagnostic(Diagnostic.Create(ParamNameRequiredRule, context.Node.GetLocation()));
+                return;
+            }
+
+            var argumentIndex = method.Parameters.IndexOf(paramNameArgument);
+            var argument = objectCreation.ArgumentList.Arguments[argumentIndex];
```

Comment:
I think I have an extension method for getting literals. So I would prefer to have slightly higher level code here:

var argumentLiteral = argument.Exception.AsLiteral();
if (argumentLiteral == null) {
  return; // not a string literal, analyzer doesn't care about other cases
}

var parent = GetEnclosingMethodBase(objectCreation);
Contract.Assert(parent != null, "Object creation node should definitely have a method parent");
// You may remove this check if this assumption is not valid

if (parameters.HasParameterOfName(literal)) {
  // first diagnostic
  return; // I guess we have some kind of a rule, that method returns when diagnostic was found
}

// this means that method doesn't have parameter that was used in the exception
string methodOrConstructor = method is ConstructorDeclarationSyntax ? "Constructor" : "Method";
context.ReportDiagnostic ...